garage/tf/models/base.py

```diff
@@ -0,0 +1,368 @@
+"""Base model classes."""
+import abc
+from collections import namedtuple
+import warnings
+
+import tensorflow as tf
+
+
+class BaseModel(abc.ABC):
+    """
+    Interface-only abstract class for models.
+
+    A Model contains the structure/configuration of a set of computation
+    graphs, or can be understood as a set of networks. Using a model
+    requires calling `build()` with given input placeholder, which can be
+    either tf.placeholder, or the output from another model. This makes
+    composition of complex models with simple models much easier.
+
+    Examples:
+        model = SimpleModel()
+        # To use a model, first create a placeholder.
+        # In the case of TensorFlow, we create a tf.placeholder.
+        input_ph = tf.placeholder(tf.float32)
+
+        # Building the model
+        output = model.build(input_ph)
+
+        # We can also pass the output of a model to another model.
+        # Here we pass the output from the above SimpleModel object.
+        model_2 = ComplexModel()
+        output_2 = model_2.build(output)
+
+    """
+
+    def build(self, *inputs):
+        """
+        Output of model with the given input placeholder(s).
+
+        This function is implemented by subclasses to create their computation
+        graphs, which will be managed by Model. Generally, subclasses should
+        implement `build()` directly.
+        """
+        pass
+
+    @property
+    def name(self):
+        """Name for this Model."""
+        pass
+
+    @property
+    def input(self):
+        """
+        Tensor input of the Model.
+
+        The input can be either a placeholder, or the output from another
+        model.
+        """
+        pass
+
+    @property
+    def output(self):
```

Comment:
Yes, I have the same thought. That's why I suggested `BaseModel` would not be needed.

Reply:
Honestly, the only part of this API which seems reasonably common across Tensorflow and PyTorch is something like `output = model.build(input)`. PyTorch would usually call that method `forward`, but adding an alias is reasonable.

Having `input` and `output` fields is also unusual / considered bad practice in PyTorch, since that makes evaluating the `Module` multiple times weird.